#1 Feature/login

channnoh opened the pull request:
### 변경사항

**AS-IS**
프로젝트 시작

**TO-BE**
* JWT, Security 인증 구현
* 회원가입
    - 회원가입 요청에 대해 validation
    -  사용자 권한의 기본값은 USER
 * 로그인
    - Email, Password를 통해 로그인 인증
* 회원탈퇴
    - MemberStatus 변경하여 회원 탈퇴 상태 관

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [x] API 테스트

channnoh committed:
chore: "Remove application.properties file"

channnoh committed:
chore: "Add application.yml file"

channnoh committed:
chore: "Add data properties to application.yml"

channnoh committed:
chore: "Add jwt properties"

channnoh committed:
feat: "Add BaseEntity"

생성일, 수정일 필드 공통처리 위한 BaseEntity 생성과 JPA의 Audit 기능 추가

channnoh committed:
feat: "Member domain 생성"

Entity, Repository, Status/Type enum 생성

channnoh committed:
feat: "CustomUserDetails, CustomUserDetailsService 생성"

channnoh committed:
feat: "Token 생성을 위한 TokenProvider 생성"

channnoh committed:
test: "Token 생성 test"

channnoh committed:
feat: "Token 사용하여 JWT 기능 구현"

channnoh committed:
feat: "SecurityConfig 생성"

channnoh committed:
refactor: "Member Entity 필드 수정"

userId -> username

channnoh committed:
feat: "GlobalExceptionHandler, CustomException 기능 구현"

Global Exception, Error 처리 구현

channnoh committed:
feat: "회원가입, 로그인 기능구현"

channnoh committed:
feat: "회원탈퇴 기능구현"

mingming-mentor commented on src/main/java/com/project/goalchallenge/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,56 @@
+  private MemberStatus memberStatus = ACTIVE;
+
+  // 탈되일
+  private LocalDateTime withdrawalDate;
```
type이 dateTime이면 필드명도 xxDateTime이면 좋을 것 같습니다

mingming-mentor commented on src/main/java/com/project/goalchallenge/domain/member/dto/SignUpDto.java:
```diff
@@ -0,0 +1,61 @@
+    private MemberType memberType;
+
+
+    public Member toEntity() {
```
entity로 만들때 memberType이 빠진것 같습니다

channnoh replied:
@mingming-mentor  MemberEntity에서 memberType의 default를 user로 설정해놔서 추가하지 않았는데 추가하는 것이 좋을까요?

mingming-mentor commented on src/main/java/com/project/goalchallenge/domain/member/service/MemberService.java:
```diff
@@ -0,0 +1,67 @@
+  private final TokenProvider tokenProvider;
+
+  public SignUpDto.Response signUp(SignUpDto.Request request) {
+    boolean exists = this.memberRepository.existsByEmail(request.getEmail());
+    if (exists) {
+      throw new MemberException(ALREADY_REGISTER_USER);
+    }
```
exists를 재활용하지 않는다면 if문안에 바로 넣어줘도 좋아요. 그러면 불필요한 변수 선언이 줄어들 것 같습니다

mingming-mentor commented on src/main/java/com/project/goalchallenge/global/auth/model/CustomUserDetails.java:
```diff
@@ -0,0 +1,50 @@
+@RequiredArgsConstructor
+public class CustomUserDetails implements UserDetails {
+
+  private final Member member;
```
CustomUserDetails에서 Member를 그대로 field로 들고 있다면 entity Manager가 CustomUserDetails가 존재하는 곳까지 있어야할 것 같아요. CustomUserDetail에서 Member type을 들고 있기 보다는 필요한 필드만 들고 있도록 변경되면 좋을것 같습니다

mingming-mentor commented on src/main/java/com/project/goalchallenge/global/config/exception/CustomException.java:
```diff
@@ -0,0 +1,17 @@
+  private final String message;
+
+  public CustomException(ErrorCode errorCode) {
+    this.errorCode = errorCode;
+    this.status = errorCode.getStatus();
+    this.message = errorCode.getMessage();
```
요렇게 설정이 되는거면 errorCode만 들고 있어도 되지 않을까하는 생각이 드네요ㅎㅎ 사용할때 getStatus, getMessage를 이용하면 되지 않을까 싶어요

mingming-mentor reviewed (approved):
리뷰드렸습니다! 수고 많으셨어요

channnoh committed:
refactor: "LocalDateTime 객체 이름 xxxDateTime 으로 변경"

channnoh committed:
refactor: "불필요한 변수 선언 삭제"

channnoh committed:
refactor: "불필요한 변수 선언 삭제"

CustomException.java 에서 int status, String message 필드 제거하고 getStatus, getMessage 사용

channnoh committed:
refactor: "Member.java toEntity() memberType 추가 memberType 필수값으로 수정"

channnoh committed:
refactor: "CustomUserDetails 필요한 field만 관리하도록 수정"

channnoh committed:
feat: "탈퇴한 유저 로그인 불가하도록 기능 추가"

channnoh committed:
refactor: "Exception log 수정"

channnoh committed:
docs: "ERD 수정"

channnoh committed:
refactor: "Global Exception 디렉토리 구조 변경"

channnoh committed:
feat: "이미 탈퇴한 회원 예외 추가"
